CONTRIBUTING.md

```diff
@@ -0,0 +1,209 @@
+- `/test-app` - A series of test CDK applications which consume `functionless` via npm. We use this to test compilation of various "real world" scenarios and CDK synthesis.
+
+Functionless leverages the [typescript compiler api](https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API). `compile.ts` runs over the ts of a consuming application, validating, transforming, and extracting necessary information.
+
+> Note: Functionless does not compile itself, at the moment, functionless like `reflect` cannot be used to transform the core functionless source. `/test`s are compiled.
+
+The compiler outputs the functionless AST (`/src/expression.ts`, `/src/declaration.ts`, `/src/statement.ts`) into the transpiled javascript which is used by the resources (`/src/function.ts`, `/src/step-function.ts`, etc) during **cdk synthesis** to infer infrastructure (CDK Constructs) and transpile typescript/Functionless AST to cloud native languages (ASL, Velocity).
+
+## Building Functionless
+
+```console
+$ yarn install
+$ yarn watch
+```
+
+### Guidelines
+
+- Every change or bug fix MUST come with one or more tests that would break before the change.
+- Errors MUST be documented as [Error Codes](#error-codes) with detailed descriptions.
+- Errors SHOULD be discovered via `/src/validate.ts`. `/src/validate.ts` runs with the `functionless` cli and the `@functionless/language-service` in the IDE to give early errors to consumers.
+- No/Minimal Magic - Core Functionless resources should improve the developer experience without introducing opinionated implementation. The goal is to translate typescript behavior into AWS services. The infrastructure created should be clear and logical. For example, `Step Functions` cannot perform arithmetic natively, we will not automatically create/call a lambda to do it. We will validate for arithmetic, throw an error, and show how a `Function` can be used as a workaround.
+
+### Project Configuration Changes
+
+Functionless uses [`projen`](https://projen.io/), desired state project definitions. Most configuration files in the main `Functionless` package are generated by project and should not be updated manually (`package.json`, `prettierrc.json`, `tsconfig.json`, etc).
+
+To make changes, update the `projen.js` file and run `yarn default`.
+
+The sub-packages (ex: website, test-app) do not currently use projen ([TODO: mono repo](https://github.com/functionless/functionless/issues/258)), their package configuration can be updated manually.
+
+### Testing
+
+Required:
+
+- Docker
+  - With the docker daemon running.
+- [Localstack](https://docs.localstack.cloud/get-started/#localstack-cli)
+
+Commands:
+
+- `yarn test` - boots up localstack, runs all tests, stops localstack
+- `yarn test:fast` - runs all non-localstack tests
+- `yarn test:update` - runs all tests and updates snapshots
+- `yarn watch` - watches all tests
+- `vs-code jest plugin` - use the vs code jest plugin to run individual tests and debug
```

Comment:
I think I need an example of this. How does it work

Reply:
Not meant to be a command, separated it.